tests/48admin.pl

```diff
@@ -40,30 +42,55 @@ sub await_purge_complete {
       # tightly control the actions taken by that user.
       # Conceivably this API may change based on the number of API calls the
       # user made, for instance.
+      print STDERR "zzdo";
       matrix_register_user( $http, "admin" )
       ->then( sub {
          ( $user ) = @_;
 
-         do_request_json_for( $user,
-            method => "GET",
-            uri    => "/v3/admin/whois/".$user->user_id,
-         )
-      })->then( sub {
-         my ( $body ) = @_;
+         print STDERR "zzmru";
 
-         assert_json_keys( $body, qw( devices user_id ) );
-         assert_eq( $body->{user_id}, $user->user_id, "user_id" );
-         assert_json_object( $body->{devices} );
+         # Synapse flushes IP addresses to the database every 5 seconds, so we
+         # need to keep checking because the IP address won't appear for a few
+         # seconds (unless the worker that flushes the IP addresses is the same
+         # as the one that handles /whois).
+         retry_until_success {
+            print STDERR "zzrus";
 
-         foreach my $value ( values %{ $body->{devices} } ) {
-            assert_json_keys( $value, "sessions" );
-            assert_json_list( $value->{sessions} );
-            assert_json_keys( $value->{sessions}[0], "connections" );
-            assert_json_list( $value->{sessions}[0]{connections} );
-            assert_json_keys( $value->{sessions}[0]{connections}[0], qw( ip last_seen user_agent ) );
-         }
+            return do_request_json_for( $user,
+               method => "GET",
+               uri    => "/v3/admin/whois/".$user->user_id,
+            )->then( sub {
+               my ( $body ) = @_;
 
-         Future->done( 1 );
+               print STDERR "zzdrjf";
+
+               assert_json_keys( $body, qw( devices user_id ) );
+               assert_eq( $body->{user_id}, $user->user_id, "user_id" );
+               assert_json_object( $body->{devices} );
+
+               foreach my $value ( values %{ $body->{devices} } ) {
+                  assert_json_keys( $value, "sessions" );
+                  assert_json_list( $value->{sessions} );
+                  assert_json_keys( $value->{sessions}[0], "connections" );
+                  assert_json_list( $value->{sessions}[0]{connections} );
+                  assert_json_object( $value->{sessions}[0]{connections}[0] );
+               }
+
+               Future->done( $body->{devices} );
+            });
+         }, initial_delay => 0.5,
```

Comment:
ugh. Turns out you need to omit this comma:

```suggestion
         } initial_delay => 0.5,
```

or, possibly less weirdly, stick a `sub` in front of the `{`.